crates/sui-graphql-client/src/error.rs

```diff
@@ -0,0 +1,316 @@
+        Self::empty().with_parse_error(ParseError::UrlParseError(error.to_string()))
+    }
+}
+
+impl From<ParseIntError> for Error {
+    fn from(_: ParseIntError) -> Self {
+        Self::empty().with_parse_error(ParseError::BigIntParsingError)
+    }
+}
+
+impl From<ParseError> for Error {
+    fn from(error: ParseError) -> Self {
+        Self::empty().with_parse_error(error)
+    }
+}
+
+impl From<AddressParseError> for Error {
+    fn from(error: AddressParseError) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl From<base64ct::Error> for Error {
+    fn from(error: base64ct::Error) -> Self {
+        Self::empty().with_deserialize_error(DeserializeError(error.to_string()))
+    }
+}
+
+impl From<chrono::ParseError> for Error {
+    fn from(error: chrono::ParseError) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl From<DigestParseError> for Error {
+    fn from(error: DigestParseError) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl From<TryFromIntError> for Error {
+    fn from(error: TryFromIntError) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl From<ConversionError> for Error {
+    fn from(error: ConversionError) -> Self {
+        Self::empty().with_conversion_error(error)
+    }
+}
+
+impl From<TypeParseError> for Error {
+    fn from(error: TypeParseError) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl From<ConflictingArguments> for Error {
+    fn from(error: ConflictingArguments) -> Self {
+        Self::empty().with_error(Box::new(error))
+    }
+}
+
+impl Error {
+    /// Create an empty error with no source.
+    fn empty() -> Self {
+        Self {
+            inner: Box::new(ClientError {
+                query_response_error: false,
+                deserialize_error: false,
+                parse_error: false,
+                empty_response_error: false,
+                conversion_error: false,
+                source: None,
+            }),
+        }
+    }
+
+    /// Convert the given error into a generic error.
+    fn from_error<E: Into<BoxError>>(error: E) -> Self {
+        let boxed_error = error.into();
+        let result = Self::empty();
+
+        if let Some(query_response_error) = boxed_error.downcast_ref::<QueryResponseError>() {
+            result.with_query_response_error(query_response_error.clone())
+        } else if let Some(deserialize_error) = boxed_error.downcast_ref::<DeserializeError>() {
+            result.with_deserialize_error(deserialize_error.clone())
+        } else if let Some(parse_error) = boxed_error.downcast_ref::<ParseError>() {
+            result.with_parse_error(parse_error.clone())
+        } else if let Some(conversion_error) = boxed_error.downcast_ref::<ConversionError>() {
+            result.with_conversion_error(conversion_error.clone())
+        } else {
+            result.with_error(boxed_error)
+        }
```

Comment:
our library code shouldn't have to do any downcasting, this should be done by users if they want to get the concrete type of the error source